Filtering through rating

saravoong opened the pull request:
Added the ability to filter through rating, also, I can fix the margin spacing if the genre filter bothers anyone
<img width="781" alt="Screenshot 2024-05-28 at 9 18 41 PM" src="https://github.com/shaaronl/jukeboxd/assets/129801695/6dce25d7-2cbb-450a-b3d7-5ffa9ed26429">

Commit:
Filtering through rating

tammy-si commented:
Your pop is cut off in your screenshot.

tammy-si commented:
No comments. All good to me

tammy-si commented on packages/react-frontend/src/routes/Album.jsx:
```diff
@@ -75,17 +108,37 @@ export default function Album() {
+            value={selectedRating}
+            onChange={(e) => setSelectedRating(e.target.value)}
+          >
+            <option value="">RATING</option>
```
Why is the value just "". Will this eventually be something?

saravoong replied:
The value is just "" because it indicates when a filter is not chosen, so it won't apply to the filtered albums. I tested it without the `value=""` parts and it doesn't work, when I apply all the filters, then unfilter, it doesn't show any albums afterwards.

shaaronl commented on packages/react-frontend/src/routes/Album.jsx:
```diff
@@ -20,6 +19,40 @@ async function fetchArtistBySpotifyId(spotifyId) {
+    const averageRatings = {};
+    for (const albumId in albumRatings) {
+      const { totalRating, count } = albumRatings[albumId];
+      averageRatings[albumId] = Math.floor(totalRating / count);
```
hmm, i see that you filter by the floored average,  but this could potentially lead to incorrect album options getting displayed. this is because when i displayed the average stars, i didn't floor it, instead i displayed the rating up to one decimal place. so i think if you filter by '3' this will lead to all options ranging from 3.1-3.9 showing up which isn't what our function should do... i could floor the avg ratings getting displayed if that would be easier though

Yakkaya replied:
I like that its floored because we would want, for example, the 3 star filter to show albums reviewed between 3 and 4 stars on average. Its more intuitive than having a 2.9 star album populate under the 3 star filter.

saravoong replied:
Yeah, I had the same thing in mind as Yasemin, I think that users might get confused if we start rounding up/down the average ratings.

tammy-si commented on packages/react-frontend/src/routes/Album.jsx:
```diff
@@ -20,6 +19,40 @@ async function fetchArtistBySpotifyId(spotifyId) {
+    const albumRatings = {};
+
+    // Calculate the total rating and count for each album
+    reviews.forEach((review) => {
+      const { album_id, rating } = review;
+      if (!albumRatings[album_id]) {
+        albumRatings[album_id] = { totalRating: 0, count: 0 };
+      }
+      albumRatings[album_id].totalRating += rating;
+      albumRatings[album_id].count += 1;
+    });
+
+    // Calculate the floored average rating for each album
+    const averageRatings = {};
+    for (const albumId in albumRatings) {
+      const { totalRating, count } = albumRatings[albumId];
+      averageRatings[albumId] = Math.floor(totalRating / count);
+    }
```
I understand what you're going for here and how it works. However it seems a bit inefficient and maybe contributing to the slowness of the filtering.

saravoong replied:
Hmmm, I agree. Do you have any ideas on how to make the code more efficient? Maybe another way to fetch the data or what?

Yakkaya replied:
I agree that there may be a more efficient way to do this. One thing that comes to mind is storing the average rating for each album in the JSON data, then updating that value with POST api each time a new review is submitted.

saravoong replied:
Ooooo, that sounds like that could work. I'll look into it, thank you for the suggestion!

saravoong replied:
I think because the filter works out, I don't think I'm going to change the code. It's also faster after I changed the album schema.

shaaronl commented on packages/react-frontend/src/routes/Album.jsx:
```diff
@@ -28,6 +61,8 @@ export default function Album() {
+  const [selectedRating, setSelectedRating] = useState("");
 
   useEffect(() => {
     fetch("http://localhost:8000/albums")
```
when you merge this, we will have to change the link from 'localhost' to 'jukeboxd-music.azurewebsites.net'

saravoong replied:
Thank you for the heads up!

tammy-si reviewed (commented):
I think overall it looks okay and it works somewhat.

shaaronl commented on packages/react-frontend/src/routes/Album.jsx:
```diff
@@ -44,28 +79,26 @@ export default function Album() {
   useEffect(() => {
     const filterAlbums = async () => {
```
the filter works, but it seems a bit slow. is there anyway we can reduce the time it takes to filter by something?

saravoong replied:
Hmmm, Yasemin made a suggestion of adding an average rating to the album data, so I'm looking into it.

shaaronl commented on packages/react-frontend/src/routes/Album.jsx:
not sure if this would be too difficult to implement, but do you think our filter should have a reset button?

saravoong replied:
Oooooo, that sounds like a great idea! I'm going to try to implement that.

BraedenAlonge commented:
Can i suggest changing "pop" to "POP" or "VIEWS"? Especially since "RATING" is capitalized.

saravoong commented:
> Your pop is cut off in your screenshot.

saravoong commented:
> Your pop is cut off in your screenshot.

Hmm yes, I'm planning on fixing it!

saravoong commented:
> Can i suggest changing "pop" to "POP" or "VIEWS"? Especially since "RATING" is capitalized.

Yeah, that seems like a great idea, so there's no need to change the CSS of it! I just followed the Figma, but I'll definitely implement that.

Yakkaya reviewed (approved):
looks good overall but I highly suggest we brainstorm a new way to get averages as a team, then work on implementing that before the end of our sprint

Commit:
Added a reset button for the filters

saravoong commented:
I added a reset button for the filters, here what it looks like. Suggestion made by @shaaronl 
<img width="528" alt="Screenshot 2024-05-30 at 12 24 08 PM" src="https://github.com/shaaronl/jukeboxd/assets/129801695/c87610b0-04dd-435f-8f1e-9720af05bf47">

Commit:
Added genres to the album schema

saravoong commented:
I added the genres (from the artist data) to the album schema. Then I updated Album.jsx with this new change, so this should contribute to helping the slowness of the filters. I still however need to update the average rating logic.

shaaronl commented:
> I added a reset button for the filters, here what it looks like. Suggestion made by @shaaronl <img alt="Screenshot 2024-05-30 at 12 24 08 PM" width="528" src="https://private-user-images.githubusercontent.com/129801695/335353383-c87610b0-04dd-435f-8f1e-9720af05bf47.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcxMjI4OTgsIm5iZiI6MTcxNzEyMjU5OCwicGF0aCI6Ii8xMjk4MDE2OTUvMzM1MzUzMzgzLWM4NzYxMGIwLTA0ZGQtNDM1Zi04ZjFlLTk3MjBhZjA1YmY0Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUzMVQwMjI5NThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYmJiNmRjM2M3MWQyZTY3YTljZDFiYmZlYjRlNWI1ODk4NDJhMmYzMWJhYTEwM2EzNDZlNDEwMDYzNTgyOGY4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bJYsoYj53PJshqZCV5AQw_i-UEVKG37QxDsW1RKqIh0">

nice! i like the button! I'm wondering if we should readjust either the filtering components/button or our logo though because the reset button looks almost as big as our logo haha. thoughts?

Commit:
I need to delete this branch

saravoong commented:
I accidentally pushed the wrong thing to this branch, but because another branch (the search bar) had this same exact code, you can ignore this pull request
